### PR TITLE
GSW-1838 refactor: token_register

### PR DESCRIPTION
**Pull request by r3v4s**
## refactor
1. replace `map` to `avl` for registered token list with its interface

2. more testcases

### CODE REVIEW

File: staker/token_register.gno [hunk 1]
```diff
@@ -21,144 +22,114 @@ type GRC20Interface interface {
...
 	return tokens
 }
 
 // RegisterGRC20Interface registers a GRC20 token interface
 func RegisterGRC20Interface(pkgPath string, igrc20 GRC20Interface) {
 	prevAddr := std.PrevRealm().Addr()
 	prevPath := std.PrevRealm().PkgPath()
-	if !(prevAddr == consts.TOKEN_REGISTER || prevPath == consts.INIT_REGISTER_PATH || strings.HasPrefix(prevPath, "gno.land/r/g1er355fkjksqpdtwmhf5penwa82p0rhqxkkyhk5")) {
+	if !(prevAddr == consts.TOKEN_REGISTER || prevPath == consts.INIT_REGISTER_PATH || strings.HasPrefix(prevPath, consts.TOKEN_REGISTER_NAMESPACE)) {
 		panic(addDetailToError(
```
Comment (dongwon8247) [inline staker/token_register.gno:46]: Avoid using panic for general errors. Return errors instead

File: staker/token_register.gno [hunk 2]
```diff
@@ -21,144 +22,114 @@ type GRC20Interface interface {
 }
 
 var (
-	registered = make(map[string]GRC20Interface)
+	registered = avl.NewTree()
 	locked     = false // mutex
```
Comment (dongwon8247) [inline staker/token_register.gno:26]: global variables?

File: staker/token_register.gno [hunk 3]
```diff
@@ -21,144 +22,114 @@ type GRC20Interface interface {
...
-	if !(prevAddr == consts.TOKEN_REGISTER || prevPath == consts.INIT_REGISTER_PATH || strings.HasPrefix(prevPath, "gno.land/r/g1er355fkjksqpdtwmhf5penwa82p0rhqxkkyhk5")) {
+	if !(prevAddr == consts.TOKEN_REGISTER || prevPath == consts.INIT_REGISTER_PATH || strings.HasPrefix(prevPath, consts.TOKEN_REGISTER_NAMESPACE)) {
 		panic(addDetailToError(
 			errNoPermission,
-			ufmt.Sprintf("token_register.gno__RegisterGRC20Interface() || only register(%s) can register token, called from %s", consts.TOKEN_REGISTER, prevAddr),
+			ufmt.Sprintf("only register(%s) can register token, called from %s", consts.TOKEN_REGISTER, prevAddr),
 		))
 	}
 
 	pkgPath = handleNative(pkgPath)
```
Comment (dongwon8247) [inline staker/token_register.gno:52]: `handleNative` is being called multiple times overall. Make sure it's called only once where only needed.

File: staker/token_register.gno [hunk 4]
```diff
@@ -170,3 +141,30 @@ func handleNative(pkgPath string) string {
 
 	return pkgPath
 }
+
+func getToken(pkgPath string) GRC20Interface {
+	pkgPath = handleNative(pkgPath)
+
+	itoken, _ := registered.Get(pkgPath)
+	return itoken.(GRC20Interface)
```
Comment (dongwon8247) [inline staker/token_register.gno:149]: There should be checks. Add type assertions with error handling.

File: staker/token_register.gno (hunk 1 above)
Comment (dongwon8247) [reply staker/token_register.gno:46]: This and dealing with global variables go the same for your other ongoing refactor work, including opened PRs. Address them.

**Review by dongwon8247 — CHANGES_REQUESTED** [review]
Is this file necessary? 
Remove all parts related to gnoswap-labs/grc20-register, replace it with https://github.com/gnolang/gno/pull/3135.

**Comment by r3v4s** [comment]
changing to draft
- new pr will be made to support grc20reg from gno
- if grc20reg works fine, this pr will be closed.

**Comment by r3v4s** [comment]
grc20reg has been merged into gnolang/gno master, and it works fine with current gnoswap. Therfore closing this.

**Pull request closed without merging**